### ISSUE TITLE
Bounded while loops in symbolic execution

### ISSUE THREAD

**Issue by benozol** [colis-anr/colis-language#56]
Currently, the symbolic execution of any while loops diverges. The symbolic execution of while loops should instead be bounded, easiest is by a maximal number of loop iterations.

**Comment by Niols** [comment]
Agreed. This would also allow to differentiate timeouts (limitation of our tool) and probably infinite loops (bugs)

**Comment by benozol** [comment]
@niols @claudemarche We should discuss the behaviour of executions that hit the maximal number of loop iterations. I guess we can reuse `BExit` which 1) is always transmitted/cannot be cought in CoLiS, but 2) it also results from the instruction `IExit` and from failing utilities outside of conditions, so we wont be able to distinguish them.

**Comment by Niols** [comment]
Yet an other return code?

**Comment by benozol** [comment]
Possibly, but I was referring to the behaviour in the semantics: 

https://github.com/colis-anr/colis-language/blob/37234312ee520c52e844251e3f27cb35b69c402b/src/concrete/semantics.mlw#L538-L543

The difficulty is that there is nothing corresponding in Shell. Open questions are:
1) should a loop boundary be cought in subshells
2) what should be the return value (note that return value and behaviour is disconnected in CoLiS semantics)

For example, what do we expect when running the following in CoLiS with bounded loops:
```shell-script
( while 1; do echo -n; done );
echo $?
```

And if we say that out-of-boundary loops in subshells should be cought, the behaviour of loop boundaries corresponds to the exit behaviour and we could conveniently reuse the latter.

**Comment by Niols** [comment]
Can't the semantics just get stuck? Or should the semantics describe also diverging programs? You only need the limit in the interpreter, right?

I'm sorry, I'm not sure I get all the aspects of this question.

**Comment by claudemarche** [comment]
I think that when the loop bound is reached, one has to raise an 
exception that is never caught : if a loop doesn't reach its end, what 
is going on afterwards does not correspond to any concrete execution, so 
symbolic execution must stop globally.

A new return code is certainly appropriate here

- Claude

Le 24/01/2019 à 17:50, Benedikt Becker a écrit :
> Possibly, but I was referring to the behaviour in the semantics:
> 
> https://github.com/colis-anr/colis-language/blob/37234312ee520c52e844251e3f27cb35b69c402b/src/concrete/semantics.mlw#L538-L543
> 
> The difficulty is that there is nothing corresponding in Shell. Open questions are:
> 1) should a loop boundary be cought in subshells
> 2) what should be the return value (note that return value and behaviour is disconnected in CoLiS semantics)
> 
> For example, what do we expect when running the following in CoLiS with bounded loops:
> ```shell-script
> ( while 1; do echo -n; done );
> echo $?
> ```
> 
> And if we say that out-of-boundary loops in subshells should be cought, the behaviour of loop boundaries corresponds to the exit behaviour and we could conveniently reuse the latter.
>

**Comment by benozol** [comment]
> when the loop bound is reached, raise an exception that is never caught

I agree, and this difference in the behaviour of subshells (the only point where exit behaviour is cought) requires a separate behaviour.

> Can't the semantics just get stuck?

If we want to look at the over-approximation property of the symbolic execution we have to model the loop boundary in the semantics, too.
